pxtblocks/fields/field_matrix.ts

```diff
@@ -0,0 +1,267 @@
+import * as Blockly from "blockly";
+
+interface MatrixDisplayProps {
+    cellWidth: number;
+    cellHeight: number;
+    cellLabel: string;
+    cellHorizontalMargin: number;
+    cellVerticalMargin: number;
+    cornerRadius: number;
+    cellStroke?: string;
+    cellFill?: string;
+    padLeft?: number;
+    scale?: number;
+}
+
+export abstract class FieldMatrix extends Blockly.Field {
+    protected cells: SVGRectElement[][] = [];
+    protected matrixSvg: SVGSVGElement;
+    private keyDownBinding: Blockly.browserEvents.Data | null = null;
+    private blurBinding: Blockly.browserEvents.Data | null = null;
+    protected selected: [x: number, y: number] | undefined = undefined;
+
+    protected abstract numMatrixCols: number;
+    protected abstract numMatrixRows: number;
+    protected abstract clearSelectionOnBlur: boolean;
+
+    protected createMatrixDisplay({
+        cellWidth,
+        cellHeight,
+        cellLabel,
+        cellHorizontalMargin,
+        cellVerticalMargin,
+        cornerRadius = 0,
+        cellFill,
+        cellStroke,
+        padLeft = 0,
+        scale = 1
+    }: MatrixDisplayProps): void {
+        // Initialize the matrix that holds the state
+        for (let i = 0; i < this.numMatrixCols; i++) {
+            this.cells.push([]);
+        }
+
+        // Create the cells of the matrix that is displayed
+        for (let y = 0; y < this.numMatrixRows; y++) {
+            const row = pxsim.svg.child(this.matrixSvg, "g", { 'role': 'row' });
+            for (let x = 0; x < this.numMatrixCols; x++) {
+                const tx = scale * x * (cellWidth + cellHorizontalMargin) + cellHorizontalMargin + padLeft;
+                const ty = scale * y * (cellHeight + cellVerticalMargin) + cellVerticalMargin;
+
+                const cellG = pxsim.svg.child(row, "g", { transform: `translate(${tx} ${ty})`, 'role': 'gridcell' });
+                const rectOptions = {
+                    'id': `${this.sourceBlock_.id}:${x}${y}`,  // For aria-activedescendant
```

Comment:
could we move this id formatting into a function to make sure the format stays consistent here and in the aria-activedescendant bit below?